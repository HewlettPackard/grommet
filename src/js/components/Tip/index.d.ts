import * as React from 'react';
// import { Omit } from "../../utils";
import { DropProps } from '../Drop';

export interface TipProps {
  content?: React.ReactNode;
  dropProps?: DropProps;
}

// TODO polish after resolving Box wrapper
// declare const Tip: React.ComponentClass<TipProps & Omit<JSX.IntrinsicElements['TODO'], 'title'>>;
declare const Tip: React.ComponentClass<TipProps>;

export { Tip };
