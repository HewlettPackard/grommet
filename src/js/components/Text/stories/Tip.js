import React from 'react';

import { grommet, Box, Text, Grommet } from 'grommet';

const alphabet = 'a b c d e f g h i j k l m n o p q r s t u v w x y z';

export const Tip = () => (
  <Grommet theme={grommet}>
    <Box align="center" pad="medium" gap="xlarge">
      <Box width="small">
        <Text tip={alphabet} truncate>
          {alphabet}
        </Text>
      </Box>
      <Text
        tip={{ dropProps: { align: { left: 'right' } }, content: 'tooltip' }}
      >
        Tip with dropProps
      </Text>
      <Text
        tip={{
          plain: true,
          dropProps: { align: { bottom: 'top' } },
          content: (
            <Box
              pad="xxsmall"
              elevation="small"
              background="#EDEDED" // no opacity
              round="xsmall"
              margin="xsmall"
              overflow="hidden"
              align="center"
            >
              tooltip
            </Box>
          ),
        }}
      >
        Tip with content prop
      </Text>
    </Box>
  </Grommet>
);

Tip.parameters = {
  chromatic: { disable: true },
};

export default {
  title: `Type/Text/Tip`,
};
