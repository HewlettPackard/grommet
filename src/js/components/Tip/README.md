## undefined
Arbitrary text.

[![](https://cdn-images-1.medium.com/fit/c/120/120/1*TD1P0HtIH9zF0UEH28zYtw.png)](https://storybook.grommet.io/?selectedKind=Tip&full=0&addons=0&stories=1&panelRight=0) [![](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/github/grommet/grommet-sandbox?initialpath=/tip&module=%2Fsrc%2FTip.js)
## Usage

```javascript
import { Tip } from 'grommet';
<Tip />
```

## Properties

**textAlign**

How to align the text inside the component. Defaults to `start`.

```
start
center
end
```
  
## Intrinsic element

```
span
```
## Theme
  
**tip**

Any valid Drop property for the Tip. Expects `object`.

Defaults to

```
undefined
```

**tip.container**

Any valid Box property for the Tip container. Expects `object`.

Defaults to

```
{ round: 'small', elevation: 'small' }
```

**global.edgeSize**

The possible sizes for margin. Expects `object`.

Defaults to

```
{
    edgeSize: {
      none: '0px',
      hair: '1px',
      xxsmall: '3px',
      xsmall: '6px',
      small: '12px',
      medium: '24px',
      large: '48px',
      xlarge: '96px',
      responsiveBreakpoint: 'small',
    },
  }
```
