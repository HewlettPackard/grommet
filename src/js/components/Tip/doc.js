import { describe, PropTypes } from 'react-desc';

import { getAvailableAtBadge } from '../../utils/mixins';

export const doc = Tip => {
  const DocumentedTip = describe(Tip)
    .availableAt(getAvailableAtBadge('Tip'))
    .description(
      `Tooltip or a hint when hovering over a UI item. 
    The tooltip will render when hovering on top of the 
    Tip child node or string.`,
    )
    .usage(
      `import { Tip } from 'grommet';
<Tip />`,
    )
    .intrinsicElement('span'); // TODO after resolving the Box wrapper issue

  DocumentedTip.propTypes = {
    content: PropTypes.node.description(`The tooltip content inside the drop.`),
    dropProps: PropTypes.object
      .description('Any valid Drop prop to style the Tip drop container.')
      .defaultValue({
        align: { top: 'bottom' },
        plain: true,
        trapFocus: false,
      }),
  };

  return DocumentedTip;
};

export const themeDoc = {
  'tip.content': {
    description: 'Any valid Box property for the Tip container.',
    type: 'object',
    defaultValue: undefined,
  },
  'tip.drop': {
    description: 'Any valid Drop property for the Tooltip.',
    type: 'object',
    defaultValue: undefined,
  },
};
