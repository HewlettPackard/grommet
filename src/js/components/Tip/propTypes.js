import PropTypes from 'prop-types';

let PropType = {};
if (process.env.NODE_ENV !== 'production') {
  PropType = {
    content: PropTypes.node,
    deactivated: PropTypes.bool,
    dropProps: PropTypes.object,
    plain: PropTypes.bool,
  };
}
export const TipPropTypes = PropType;
