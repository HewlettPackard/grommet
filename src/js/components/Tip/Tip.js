import React, {
  Children,
  cloneElement,
  forwardRef,
  useContext,
  useState,
  useEffect,
} from 'react';
import { ThemeContext } from 'styled-components';

import { Box } from '../Box';
import { Drop } from '../Drop';
import { useForwardedRef } from '../../utils/refs';
import { TipPropTypes } from './propTypes';

const Tip = forwardRef(({ children, content, dropProps, plain }, tipRef) => {
  const theme = useContext(ThemeContext);
  const [over, setOver] = useState(false);
  const [usingKeyboard, setUsingKeyboard] = useState();

  const onMouseDown = () => setUsingKeyboard(false);
  const onKeyDown = () => setUsingKeyboard(true);
  useEffect(() => {
    document.addEventListener('mousedown', onMouseDown);
    document.addEventListener('keydown', onKeyDown);
    return () => {
      document.removeEventListener('mousedown', onMouseDown);
      document.removeEventListener('keydown', onKeyDown);
    };
  }, []);

  const componentRef = useForwardedRef(tipRef);

  // Three use case for children
  // 1. Tip has a single child + it is a React Element => Great!
  // 2. Tip has a single child +  not React Element =>
  // span will wrap the child so we can use ref and events.
  // 3. Tip has more than one child => Abort, display Children.only error
  const child =
    (Children.count(children) <= 1 && !React.isValidElement(children) && (
      <span>{children}</span>
    )) ||
    Children.only(children);

  const clonedChild = cloneElement(child, {
    onMouseEnter: () => setOver(true),
    onMouseLeave: () => setOver(false),
    onFocus: () => {
      if (usingKeyboard) setOver(true);
    },
    onBlur: () => {
      if (usingKeyboard) setOver(false);
    },
    key: 'tip-child',
    ref: (node) => {
      // https://github.com/facebook/react/issues/8873#issuecomment-287873307
      if (typeof componentRef === 'function') {
        componentRef(node);
      } else if (componentRef) {
        // eslint-disable-next-line no-param-reassign
        componentRef.current = node;
      }
      // Call the original ref, if any
      const { ref: callerRef } = child;
      if (typeof callerRef === 'function') {
        callerRef(node);
      } else if (callerRef) {
        callerRef.current = node;
      }
    },
  });

  return [
    clonedChild,
    over && (
      <Drop
        target={componentRef.current}
        trapFocus={false}
        key="tip-drop"
        {...theme.tip.drop}
        {...dropProps}
      >
        {plain ? content : <Box {...theme.tip.content}>{content}</Box>}
      </Drop>
    ),
  ];
});

Tip.displayName = 'Tip';
Tip.propTypes = TipPropTypes;

export { Tip };
