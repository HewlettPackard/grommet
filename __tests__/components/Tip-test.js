// (C) Copyright 2014-2016 Hewlett Packard Enterprise Development LP

import React from 'react';
import renderer from 'react/lib/ReactTestRenderer';

import Tip from '../../src/js/components/Tip';

// needed because this:
// https://github.com/facebook/jest/issues/1353
jest.mock('react-dom');

describe('SkipLinks', () => {
  it('has correct default options', () => {
    const component = renderer.create(
      <div>
        <div id="test-target"></div>
        <Tip onClose={(e) => e} target="test-target" />
      </div>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders with a different colorIndex value', () => {
    const component = renderer.create(
      <div>
        <div id="test-target"></div>
        <Tip colorIndex="brand"
          onClose={(e) => e} target="test-target" />
      </div>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
