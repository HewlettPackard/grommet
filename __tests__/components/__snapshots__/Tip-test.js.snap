exports[`SkipLinks has correct default options 1`] = `
<div>
  <div
    id="test-target" />
  <span />
</div>
`;

exports[`SkipLinks renders with a different colorIndex value 1`] = `
<div>
  <div
    id="test-target" />
  <span />
</div>
`;
